pipeline/plugin_runners.go

```diff
@@ -279,59 +288,77 @@ func (ir *iRunner) Starter(h PluginHelper, wg *sync.WaitGroup) {
 	}
 
 	for !globals.IsShuttingDown() {
+
 		// ir.Input().Run() shouldn't return unless error or shutdown.
 		err := ir.input.Run(ir, h)
-		if err != nil {
+		registered, ok := ir.pConfig.InputRunners[ir.name]
+
+		if !ok || registered != ir || globals.IsShuttingDown() {
+			// Plugin was removed deliberately from the list of InputRunners or
+			// has been superseded by another instance, or we're in shutdown.
+			// In this case, avoid triggering a Heka shutdown ourselves.
+			ir.Unregister(ir.pConfig)
+			return
+		} else if err == nil {
+			// Plugin exited cleanly.
+			break
+		} else {
+			// Plugin exited by returning an error.
 			ir.LogError(err)
-		} else if !ir.transient {
-			rh.Reset()
-			ir.LogMessage("stopped")
-		}
 
-		// Send shutdown signal to any decoders that need it.
-		if len(ir.shutdownWanters) > 0 {
-			ir.shutdownLock.Lock()
-			for _, wanter := range ir.shutdownWanters {
-				wanter.Shutdown()
+			// If we don't support restart, just stop here.
+			recon, ok := ir.plugin.(Restarting)
+			if !ok {
+				break
 			}
-			ir.shutdownLock.Unlock()
-		}
 
-		// Are we supposed to stop? Save ourselves some time by exiting now.
-		if globals.IsShuttingDown() {
-			break
-		}
+			// Otherwise we'll execute the Retry config
+			recon.CleanupForRestart()
+			if ir.maker == nil {
+				ir.maker = ir.pConfig.makers["Input"][ir.name]
+			}
 
-		// If we're transient we just exit.
-		if ir.transient {
-			break
-		}
-		// We're not transient, we either try to restart or we shut down
-		// altogether.
-		recon, ok := ir.plugin.(Restarting)
-		if !ok {
-			ir.LogMessage("has stopped, shutting down.")
-			globals.ShutDown()
-			break
-		}
+		initLoop:
```

Comment:
And another bug that you've brought in from the foRunner code... we should be checking for globals.IsShuttingDown() in the initLoop, otherwise we might end up stuck here forever.